### ISSUE TITLE
How to calculate the parameters and MAGs with the torchsparse lib?

### ISSUE THREAD

**Issue by dapeng2017** [mit-han-lab/torchsparse#9]
I want to count the MAGs when running spvnas then i try the thop lib but it doesn’t work for torchsparse...So, can you tell me how to calculate them?

**Comment by kentangSJTU** [comment]
Hello,

The MACs for 3D sparse convolution layers are calculated as **kernel_map_size * in_channels * out_channels**, so you may want to use the `custom_op` argument in thop to make it work.

P.S.: In our paper we actually pre-calculated the average kernel map sizes of our model on the SemanticKITTI dataset and directly do the \#MACs computation without using thop.


Best,
Haotian

**Comment by zhijian-liu** [comment]
I'm closing this issue due to the inactivity. Please feel free to reopen it if the problem has not been resolved.

**Comment by shihongf** [comment]
Hi,

Thanks for the great work. I currently have trained a my own model and want to correctly calculate the MACs of it. Could you give me a quick guide? For example, where should I find the kernel_map_size of each layer?

Thanks in advance.

**Comment by zhijian-liu** [comment]
After the model inference, this will be stored in the `kmaps` field of the input tensor: see https://github.com/mit-han-lab/torchsparse/blob/master/torchsparse/nn/functional/conv.py#L174-L176.

**Comment by shihongf** [comment]
Thanks! But the actual result did not quite match the one reported here: https://github.com/mit-han-lab/spvnas#semantickitti-1. For example, the GMACs for SemanticKITTI_val_MinkUNet@29GMACs I calculated is almost twice than what is reported. The way I calculate the average kernel map for each layer is first do a summation of `nbsizes`, which the size of active kernel maps and then average over the entire SemanticKITTI data set. Then I use calculate the GMACs for each layer by `kernel_map_size * in_channels * out_channels`. I get the final GMACs for the model by sum them all up. Did I misunderstand anything?

**Comment by kentangSJTU** [comment]
Hi @shihongf,

Have you tried to calculate the GMACs for the full model? The 29GMACs MinkUNet model is just a channel x 0.5 version of the full model, so it should have around 1/4 GMACs comparing with the full model.

Besides, we are averaging the GMACs on the sequence 08 of SemanticKITTI (the official validation set).

I can also provide you with a FLOPs counting function for thop, which might be helpful:

```python
def count_sparseConv(
    m: torchsparse.nn.Conv3d,
    x: (torchsparse.SparseTensor,),
    y: torchsparse.SparseTensor
):
    x = x[0]
    kernel_size = make_ntuple(m.kernel_size, 3)
    os = x.s if not m.transposed else y.s
    os = make_ntuple(os, 3)
    s = make_ntuple(m.stride, 3)
    d = make_ntuple(m.dilation, 3)
    ic, oc = m.in_channels, m.out_channels
    if kernel_size != (1, 1, 1):
        kmaps = y.kmaps
        kmaps.update(x.kmaps)
        kmap_key = (os, kernel_size, s, d)
        kmap_size = kmaps[kmap_key][0].shape[0]
        total_ops = ic * oc * kmap_size
    else:
        # FC
        total_ops = ic * oc * x.F.shape[0]

    m.total_ops += torch.DoubleTensor([int(total_ops)])
```

You can use the function in the following way:

```python
macs, _ = thop.profile(net, inputs=(x,), verbose=False,
    custom_ops={
        torchsparse.nn.Conv3d: count_sparseConv
    }
)
```

The function should be returning slightly larger number than what we reported in the paper, since it also takes BN layers into consideration (while we don't because BN layers can be easily folded during inference).


Best,
Haotian

**Comment by shihongf** [comment]
Thank you for your quick reply, the script really helps!

**Comment by Provemj** [comment]
I'm sorry, but what is kmaps = y.kmaps; I couldn't find kmaps in vars(y)
